### ISSUE TITLE
Compilation problems

### ISSUE THREAD

**Issue by justneedanacc** [KolasKarpenko/GenICamImpl#3]
Hi,

I wanted to test your lib but the Promise.h part does not compile under Linux with gcc11 and Cmake 3.20.3
Are you planing on fixing this or can you tell me a way to compile this?

```
Consolidate compiler generated dependencies of target GevDeviceLib
[ 12%] Building CXX object GevDeviceLib/CMakeFiles/GevDeviceLib.dir/CDevice.cpp.o
In file included from /home/devel/GenICamImpl/GevDeviceLib/include/CDevice.h:3,
                 from /home/devel/GenICamImpl/GevDeviceLib/CDevice.cpp:1:
/home/devel/GenICamImpl/Promise/include/Promise.h:297:52: error: expected nested-name-specifier before ‘TError’
  297 |                 virtual void Reject(const typename TError& error) override
      |                                                    ^~~~~~
/home/devel/GenICamImpl/Promise/include/Promise.h:297:52: error: declaration of ‘const int TError’ shadows template parameter
  297 |                 virtual void Reject(const typename TError& error) override
      |                                     ~~~~~~~~~~~~~~~^~~~~~
In file included from /home/devel/GenICamImpl/GevDeviceLib/include/CDevice.h:3,
                 from /home/devel/GenICamImpl/GevDeviceLib/CDevice.cpp:1:
/home/devel/GenICamImpl/Promise/include/Promise.h:273:36: note: template parameter ‘TError’ declared here
  273 |         template<typename TResult, typename TError>
      |                                    ^~~~~~~~
In file included from /home/devel/GenICamImpl/GevDeviceLib/include/CDevice.h:3,
                 from /home/devel/GenICamImpl/GevDeviceLib/CDevice.cpp:1:
/home/devel/GenICamImpl/Promise/include/Promise.h:297:58: error: expected ‘,’ or ‘...’ before ‘&’ token
  297 |                 virtual void Reject(const typename TError& error) override
      |                                                          ^
/home/devel/GenICamImpl/Promise/include/Promise.h: In member function ‘virtual void PromiseContext::PromiseBase<TResult, TError>::Reject(int)’:
/home/devel/GenICamImpl/Promise/include/Promise.h:297:52: error: declaration of ‘const int TError’ shadows template parameter
  297 |                 virtual void Reject(const typename TError& error) override
      |                                     ~~~~~~~~~~~~~~~^~~~~~
In file included from /home/devel/GenICamImpl/GevDeviceLib/include/CDevice.h:3,
                 from /home/devel/GenICamImpl/GevDeviceLib/CDevice.cpp:1:
/home/devel/GenICamImpl/Promise/include/Promise.h:273:36: note: template parameter ‘TError’ declared here
  273 |         template<typename TResult, typename TError>
      |                                    ^~~~~~~~
In file included from /home/devel/GenICamImpl/GevDeviceLib/include/CDevice.h:3,
                 from /home/devel/GenICamImpl/GevDeviceLib/CDevice.cpp:1:
/home/devel/GenICamImpl/Promise/include/Promise.h:299:49: error: type/value mismatch at argument 2 in template parameter list for ‘template<class TResult, class TError> class TPromise’
  299 |                         TPromise<TResult, TError>::Reject(error);
      |                                                 ^
/home/devel/GenICamImpl/Promise/include/Promise.h:299:49: note:   expected a type, got ‘TError’
/home/devel/GenICamImpl/Promise/include/Promise.h:299:59: error: ‘error’ was not declared in this scope; did you mean ‘perror’?
  299 |                         TPromise<TResult, TError>::Reject(error);
      |                                                           ^~~~~
      |                                                           perror
/home/devel/GenICamImpl/Promise/include/Promise.h: In member function ‘virtual void PromiseContext::Promise<TResult, TError>::Reset()’:
/home/devel/GenICamImpl/Promise/include/Promise.h:321:55: error: expected nested-name-specifier before ‘TError’
  321 |                                 [this](const typename TError& error) { PromiseBase<TResult, TError>::Reject(error); },
      |                                                       ^~~~~~
/home/devel/GenICamImpl/Promise/include/Promise.h:321:55: error: declaration of ‘const int TError’ shadows template parameter
  321 |                                 [this](const typename TError& error) { PromiseBase<TResult, TError>::Reject(error); },
      |                                        ~~~~~~~~~~~~~~~^~~~~~
/home/devel/GenICamImpl/Promise/include/Promise.h:308:36: note: template parameter ‘TError’ declared here
  308 |         template<typename TResult, typename TError>
      |                                    ^~~~~~~~
/home/devel/GenICamImpl/Promise/include/Promise.h:321:61: error: expected ‘,’ or ‘...’ before ‘&’ token
  321 |                                 [this](const typename TError& error) { PromiseBase<TResult, TError>::Reject(error); },
      |                                                             ^
/home/devel/GenICamImpl/Promise/include/Promise.h: In lambda function:
/home/devel/GenICamImpl/Promise/include/Promise.h:321:55: error: declaration of ‘const int TError’ shadows template parameter
  321 |                                 [this](const typename TError& error) { PromiseBase<TResult, TError>::Reject(error); },
      |                                        ~~~~~~~~~~~~~~~^~~~~~
/home/devel/GenICamImpl/Promise/include/Promise.h:308:36: note: template parameter ‘TError’ declared here
  308 |         template<typename TResult, typename TError>
      |                                    ^~~~~~~~
/home/devel/GenICamImpl/Promise/include/Promise.h:321:99: error: type/value mismatch at argument 2 in template parameter list for ‘template<class TResult, class TError> class PromiseContext::PromiseBase’
  321 |                                 [this](const typename TError& error) { PromiseBase<TResult, TError>::Reject(error); },
      |                                                                                                   ^
/home/devel/GenICamImpl/Promise/include/Promise.h:321:99: note:   expected a type, got ‘TError’
/home/devel/GenICamImpl/Promise/include/Promise.h:321:109: error: ‘error’ was not declared in this scope; did you mean ‘perror’?
  321 |                                 [this](const typename TError& error) { PromiseBase<TResult, TError>::Reject(error); },
      |                                                                                                             ^~~~~
      |                                                                                                             perror
/home/devel/GenICamImpl/Promise/include/Promise.h: In instantiation of ‘class PromiseContext::PromiseBase<std::vector<gevdevice::UdpPort::Connection>, std::__cxx11::basic_string<char> >’:
/home/devel/GenICamImpl/Promise/include/Promise.h:329:8:   required from ‘class PromiseContext::AsyncPromise<std::vector<gevdevice::UdpPort::Connection>, std::__cxx11::basic_string<char> >’
/home/devel/GenICamImpl/Promise/include/Promise.h:415:38:   required from ‘std::shared_ptr<TPromise<TResult, TError> > PromiseContext::CreateAsync(const std::function<void(const std::function<void(const TResult&)>&, const std::function<void(const TError&)>&, const std::function<void(int)>&, std::function<bool()>)>&) [with TResult = std::vector<gevdevice::UdpPort::Connection>; TError = std::__cxx11::basic_string<char>]’
/home/devel/GenICamImpl/GevDeviceLib/CDevice.cpp:236:79:   required from here
/home/devel/GenICamImpl/Promise/include/Promise.h:297:30: error: ‘void PromiseContext::PromiseBase<TResult, TError>::Reject(int) [with TResult = std::vector<gevdevice::UdpPort::Connection>; TError = std::__cxx11::basic_string<char>]’ marked ‘override’, but does not override
  297 |                 virtual void Reject(const typename TError& error) override
      |                              ^~~~~~
/home/devel/GenICamImpl/Promise/include/Promise.h: In instantiation of ‘class PromiseContext::PromiseBase<std::vector<unsigned char>, std::__cxx11::basic_string<char> >’:
/home/devel/GenICamImpl/Promise/include/Promise.h:309:8:   required from ‘class PromiseContext::Promise<std::vector<unsigned char>, std::__cxx11::basic_string<char> >’
/home/devel/GenICamImpl/Promise/include/Promise.h:400:33:   required from ‘std::shared_ptr<TPromise<TResult, TError> > PromiseContext::Create(const std::function<void(const std::function<void(const TResult&)>&, const std::function<void(const TError&)>&, const std::function<void(int)>&, const std::function<bool()>&)>&) [with TResult = std::vector<unsigned char>; TError = std::__cxx11::basic_string<char>]’
/home/devel/GenICamImpl/GevDeviceLib/CDevice.cpp:399:62:   required from here
/home/devel/GenICamImpl/Promise/include/Promise.h:297:30: error: ‘void PromiseContext::PromiseBase<TResult, TError>::Reject(int) [with TResult = std::vector<unsigned char>; TError = std::__cxx11::basic_string<char>]’ marked ‘override’, but does not override
/home/devel/GenICamImpl/Promise/include/Promise.h: In instantiation of ‘class PromiseContext::PromiseBase<bool, std::__cxx11::basic_string<char> >’:
/home/devel/GenICamImpl/Promise/include/Promise.h:309:8:   required from ‘class PromiseContext::Promise<bool, std::__cxx11::basic_string<char> >’
/home/devel/GenICamImpl/Promise/include/Promise.h:400:33:   required from ‘std::shared_ptr<TPromise<TResult, TError> > PromiseContext::Create(const std::function<void(const std::function<void(const TResult&)>&, const std::function<void(const TError&)>&, const std::function<void(int)>&, const std::function<bool()>&)>&) [with TResult = bool; TError = std::__cxx11::basic_string<char>]’
/home/devel/GenICamImpl/GevDeviceLib/CDevice.cpp:424:46:   required from here
/home/devel/GenICamImpl/Promise/include/Promise.h:297:30: error: ‘void PromiseContext::PromiseBase<TResult, TError>::Reject(int) [with TResult = bool; TError = std::__cxx11::basic_string<char>]’ marked ‘override’, but does not override
/home/devel/GenICamImpl/Promise/include/Promise.h: In instantiation of ‘class PromiseContext::PromiseBase<std::__cxx11::basic_string<char>, std::__cxx11::basic_string<char> >’:
/home/devel/GenICamImpl/Promise/include/Promise.h:329:8:   required from ‘class PromiseContext::AsyncPromise<std::__cxx11::basic_string<char>, std::__cxx11::basic_string<char> >’
/home/devel/GenICamImpl/Promise/include/Promise.h:415:38:   required from ‘std::shared_ptr<TPromise<TResult, TError> > PromiseContext::CreateAsync(const std::function<void(const std::function<void(const TResult&)>&, const std::function<void(const TError&)>&, const std::function<void(int)>&, std::function<bool()>)>&) [with TResult = std::__cxx11::basic_string<char>; TError = std::__cxx11::basic_string<char>]’
/home/devel/GenICamImpl/GevDeviceLib/CDevice.cpp:693:58:   required from here
/home/devel/GenICamImpl/Promise/include/Promise.h:297:30: error: ‘void PromiseContext::PromiseBase<TResult, TError>::Reject(int) [with TResult = std::__cxx11::basic_string<char>; TError = std::__cxx11::basic_string<char>]’ marked ‘override’, but does not override
make[2]: *** [GevDeviceLib/CMakeFiles/GevDeviceLib.dir/build.make:76: GevDeviceLib/CMakeFiles/GevDeviceLib.dir/CDevice.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:152: GevDeviceLib/CMakeFiles/GevDeviceLib.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

**Comment by justneedanacc** [comment]
This fixes the compile problems:

```
diff --git a/GevDeviceLib/include/FrameData.h b/GevDeviceLib/include/FrameData.h
index 8bf5103..d6a1385 100644
--- a/GevDeviceLib/include/FrameData.h
+++ b/GevDeviceLib/include/FrameData.h
@@ -2,6 +2,7 @@
 
 #include <vector>
 #include <cstdint>
+#include <stddef.h>
 
 namespace gevdevice{
 
@@ -65,4 +66,4 @@ private:
 	uint32_t m_lastPacketId;
 };
 
-}
\ No newline at end of file
+}
diff --git a/Promise/include/Promise.h b/Promise/include/Promise.h
index 7b876da..516237e 100644
--- a/Promise/include/Promise.h
+++ b/Promise/include/Promise.h
@@ -294,7 +294,7 @@ private:
 			m_context.PopPool(m_id);
 		}
 
-		virtual void Reject(const typename TError& error) override
+        virtual void Reject(const TError& error) override
 		{
 			TPromise<TResult, TError>::Reject(error);
 			m_context.PopPool(m_id);
@@ -318,7 +318,7 @@ private:
 
 			TPromise<TResult, TError>::Run(
 				[this](const TResult& result) { PromiseBase<TResult, TError>::Resolve(result); },
-				[this](const typename TError& error) { PromiseBase<TResult, TError>::Reject(error); },
+                [this](const TError& error) { PromiseBase<TResult, TError>::Reject(error); },
 				[this](int progress) { PromiseBase<TResult, TError>::Progress(progress); },
 				[this]() { return PromiseBase<TResult, TError>::IsCanceled(); }
 			);
```

**Comment by KolasKarpenko** [comment]
Hi Peter! I have changed this promise implementation to std::promise and I have some fixes made, soon it will published :) Regards Nikolai 01.06.2021, 16:32, "Peter.M" ***@***.***>: Closed #3.—You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or unsubscribe.
